feat: Added async overload for UploadFiles

linkdotnet opened the pull request:
This PR will add another async overload (which should become the one and only version for `>v2`): `UploadFilesAsync`.

egil commented:
During a test, is there any condition where the task does not complete immediately?

linkdotnet commented:
> During a test, is there any condition where the task does not complete immediately?

Hmm probably not. That is why we also never saw `UploadFiles` failing. If you don't think this gives us much benefit I can revert the tests (or remove the PR completely - as this holds true for our users as well).

egil commented on CHANGELOG.md:
```diff
@@ -6,6 +6,9 @@ All notable changes to **bUnit** will be documented in this file. The project ad
 ## [Unreleased]
 
+### Added
+- Added `UploadFilesAsync` as awaitable version of `UploadFiles`. By [@linkdotnet](https://github.com/linkdotnet).
```
```suggestion
- Added awaitable `UploadFilesAsync` helper method that makes it easy to upload files to a `<InputFile>` component during testing. By [@linkdotnet](https://github.com/linkdotnet).
```

egil commented on src/bunit.web/Extensions/InputFile/InputFileExtensions.cs:
```diff
@@ -16,13 +16,30 @@ public static class InputFileExtensions
 	public static void UploadFiles(
 		this IRenderedComponent<InputFile> inputFileComponent,
 		params InputFileContent[] files)
+		=> UploadFilesAsync(inputFileComponent, files).GetAwaiter().GetResult();
```
```suggestion
		{
			var uploadTask = UploadFilesAsync(inputFileComponent, files);
			if(!uploadTask.IsCompleted)
			{
				throw new SomeGoodExceptionName("Upload did not complete synchronously. Use `UploadFilesAsync` instead.");
			}
		}
```

This is just hacked together here, but I kind of like this pattern instead of using GetAwaiter().GetResult(). We are basically warning users that they their upload did not complete, and they should use the async version. This makes the change very discoverable, and those that don't need to use the async version, they can continue to use the other version without problems.

linkdotnet replied:
Good point. At least it's better than my current anti-pattern *sync over async* to just call `GetAwaiter().GetResult()`.

Still it feels a bit strange, that the sync version doe semantically something else than the async version, not?
Anyway let's give it a shot

egil replied:
Yeah, it's a bit weird, but a good middle ground, and I prefer it to use the obsolete attribute. The key is to make the exception message very clear and obvious.

egil commented on CHANGELOG.md:
```diff
@@ -6,6 +6,9 @@ All notable changes to **bUnit** will be documented in this file. The project ad
 ## [Unreleased]
 
+### Added
+ - Added awaitable `UploadFilesAsync` helper method that makes it easy to upload files to a `<InputFile>` component during testing. By [@linkdotnet](https://github.com/linkdotnet).
```
```suggestion
 - Added the `UploadFilesAsync` extension method to `IRenderedComponent<InputFile>`, which makes it easy to upload files to a `<InputFile>` component during testing and wait for asynchronous processing of the uploaded files to complete before continuing the test. By [@linkdotnet](https://github.com/linkdotnet).
```

Lets add a bit more detail. This is what the async method enables, right?

egil commented on src/bunit.web/Extensions/InputFile/InputFileExtensions.cs:
```diff
@@ -16,13 +16,36 @@ public static class InputFileExtensions
 		params InputFileContent[] files)
+		{
+			var uploadTask = UploadFilesAsync(inputFileComponent, files);
+			if(!uploadTask.IsCompleted)
```
Do we have a test that checks this code path?

linkdotnet replied:
No, I was (and still am) a bit busy and will do that later on and will notice you.

egil reviewed (commented):
Think we are more or less there.

linkdotnet commented:
I also added sync/async tests here. I added the exception including some documentation.
As the other PR: Do we want to have it `internal` or `public`?

linkdotnet commented:
Note for myself: test for exception case

egil commented:
> I also added sync/async tests here. I added the exception including some documentation.
> As the other PR: Do we want to have it `internal` or `public`?

Let's start internal for now.

linkdotnet commented:
> > I also added sync/async tests here. I added the exception including some documentation.
> > As the other PR: Do we want to have it `internal` or `public`?
> 
> Let's start internal for now.

Hmmm, does it makes sense then to have that part here?:
```csharp
if (!uploadTask.IsCompleted)
{
	throw new TaskNotCompletedException($"Upload did not complete synchronously. Use `{nameof(UploadFilesAsync)}` instead.");
}
```

The user can't use the `async` version of the API. So I guess for now, we should remove this, if we don't want to make `UploadFilesAsync` public.

egil commented on CHANGELOG.md:
```diff
@@ -6,9 +6,15 @@ All notable changes to **bUnit** will be documented in this file. The project ad
 ### Fixed
  - The created HTML contained encoded strings. Reported by [@tobiasbrandstaedter](https://github.com/tobiasbrandstaedter). Fixed by [@linkdotnet](https://github.com/linkdotnet).
 
+### Added
```
This shows up twice. Looked like a merger failure.

linkdotnet replied:
Yes that's on me. Fixed and added the missing test.

egil reviewed (approved):
LGTM

linkdotnet commented:
I'll close this in favor of pushing `v2` and putting this feature in there.
